### PR TITLE
Configure gas overhead IGP

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Instruments `multisigIsm.verify` gas overhead for various permutations of threshold and validator set size
- Add deployment tooling for IGP gas overhead

### Drive-by

- Update GasOverheadIgp to use immutable IGP

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1299

### Backward compatibility

Yes

### Testing

Unit Tests

### CODE REVIEW

File: solidity/test/isms/multisigIsm.test.ts [hunk 1]
```diff
@@ -367,31 +367,50 @@ describe('MultisigIsm', async () => {
     });
   });
 
+  // let gasOverhead: Record<number, Record<number, number>> = {};
```
Comment (nambrot) [inline solidity/test/isms/multisigIsm.test.ts:370]: Write a comment about how to use this

File: solidity/contracts/isms/MultisigIsm.sol [hunk 2]
```diff
@@ -341,6 +343,9 @@ contract MultisigIsm is IMultisigIsm, Ownable {
             // Fail if we never found a match
             require(_validatorIndex < _validatorCount, "!threshold");
             ++_validatorIndex;
+            unchecked {
```
Comment (nambrot) [inline solidity/contracts/isms/MultisigIsm.sol:346]: Undo

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 3]
```diff
@@ -42,21 +41,56 @@ export class HyperlaneCoreDeployer<
...
       'interchainGasPaymaster',
       [],
       proxyAdmin,
-      [],
+      [owner],
+      deployOpts,
+    );
+    const interchainGasOverhead = await this.deployProxiedContract(
+      chain,
+      'interchainGasOverhead',
```
Comment (nambrot) [inline typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:60]: Didn't @tkporter want to have the overhead contract to be at the existing IGP address?

File: solidity/contracts/igps/InterchainGasPaymaster.sol [hunk 4]
```diff
@@ -29,13 +29,15 @@ contract InterchainGasPaymaster is IInterchainGasPaymaster, OwnableUpgradeable {
...
     constructor() {
-        initialize(); // allows contract to be used without proxying
+        _disableInitializers();
+        _transferOwnership(msg.sender);
     }
 
     // ============ External Functions ============
 
-    function initialize() public initializer {
+    function initialize(address owner) public initializer {
```
Comment (nambrot) [inline solidity/contracts/igps/InterchainGasPaymaster.sol:38]: Agreed that we should do this, but maybe do this wholesale in a separate PR?

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 5]
```diff
@@ -42,21 +41,56 @@ export class HyperlaneCoreDeployer<
...
+    const chainConnection = this.multiProvider.getChainConnection(chain);
+    const remotes = this.multiProvider
+      .intersect(configChains, false)
+      .multiProvider.remoteChains(chain);
+
+    const configs = remotes.map((remote) => {
+      const ismConfig = this.configMap[remote].multisigIsm;
+      const verifyCost =
+        // @ts-ignore
+        multisigIsmVerifyCosts[ismConfig.validators.length][
```
Comment (nambrot) [inline typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:77]: Throw an error if the value doesnt exist

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts (hunk 3 above)
Comment (tkporter) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:60]: Yeah, imo this would be the smoothest transition for users

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 6]
```diff
@@ -173,35 +210,17 @@ export class HyperlaneCoreDeployer<
       multisigIsm.address,
       proxyAdmin,
     );
-    // Mailbox ownership is transferred upon initialization.
-    const ownables: Ownable[] = [multisigIsm, proxyAdmin];
```
Comment (tkporter) [inline typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:177]: I don't think I understand why we don't need to transfer ownership for the multisigIsm?

File: solidity/contracts/isms/MultisigIsm.sol [hunk 7]
```diff
@@ -287,10 +288,10 @@ contract MultisigIsm is IMultisigIsm, Ownable {
      * @param _metadata ABI encoded module metadata (see MultisigIsmMetadata.sol)
      * @param _message Formatted Hyperlane message (see Message.sol).
      */
-    function _verifyMerkleProof(
+    function verifyMerkleProof(
```
Comment (tkporter) [inline solidity/contracts/isms/MultisigIsm.sol:291]: also wondering what the motivation was for making these public

File: solidity/test/isms/multisigIsm.test.ts (hunk 1 above)
Comment (tkporter) [reply solidity/test/isms/multisigIsm.test.ts:370]: Wondering if we should implement this as a hardhat script instead of putting it into tests?

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts (hunk 3 above)
Comment (tkporter) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:60]: I also wonder if we should rename the deployed overhead contract to be something like "multisigIsmIgp" and then to have the base / inner IGP be called something like "baseIgp". So that it's very clear what the use of each IGP is

**Review by tkporter — COMMENTED** [review]
great job 🙏 really neat to see the gas benchmarks too

I'll bring this up in parking lot but I'm wondering if we should consider mailbox costs in the multisig ISM IGP or if that should live in the base IGP. If it lives in the base IGP then it makes composing IGPs much easier, but it also increases gas / complexity

File: solidity/test/isms/multisigIsm.test.ts (hunk 1 above)
Comment (yorhodes) [reply solidity/test/isms/multisigIsm.test.ts:370]: probably should but this would take a while to get scaffolded

File: solidity/contracts/isms/MultisigIsm.sol [hunk 8]
```diff
@@ -203,8 +204,8 @@ contract MultisigIsm is IMultisigIsm, Ownable {
         view
         returns (bool)
     {
-        require(_verifyMerkleProof(_metadata, _message), "!merkle");
-        require(_verifyValidatorSignatures(_metadata, _message), "!sigs");
+        require(verifyValidatorSignatures(_metadata, _message), "!sigs");
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/isms/MultisigIsm.sol:207]: will restore

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts (hunk 6 above)
Comment (yorhodes) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:177]: because we do it in the constructor

File: solidity/contracts/isms/MultisigIsm.sol (hunk 7 above)
Comment (yorhodes) [reply solidity/contracts/isms/MultisigIsm.sol:291]: was part of benchmarking
will restore
